### ISSUE TITLE
File Formats: Current and Future

### ISSUE THREAD

**Issue by FlynnOConnell** [MillerBrainObservatory/LBM-CaImAn-MATLAB#7]
Edit: Moved to [discussions](https://github.com/orgs/MillerBrainObservatory/discussions)

TL;DR: For now, force data that may need to be updated independently, e.g. a 3D time-series, to be stored in a separate file. As of now, that file-type is limited to `h5`. These can be easily opened in python.

# Status

Previous branches of this pipeline used [savefast](https://www.mathworks.com/matlabcentral/fileexchange/39721-save-mat-files-more-quickly) to save `.mat` files. This is essentially obsolete as of MATLAB 2020, as you can use `-v7.3` along with `-nocompression` in the build-in `save` function. 

Additionally, allowing our datasets to play nicely between MATLAB and python requires careful handling of the file types. Several popular python libraries don't work with `.mat` files. 

We have a few obvious file-format options:

- `.mat` [scipy.io.loadmat](https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.loadmat.html): doesn't work with `-v7.3 .mat` files, which is what previous iterations of this pipeline have used. However, these `.matfiles` are 
- `.h5` [h5py](https://www.h5py.org/) is the most up to date and widely used library with interfaces in every language
- `.zarr` [zarr](https://zarr.readthedocs.io/en/stable/) has no `MATLAB` api yet, but it should be on the radar if one is ever developed in the future. 

# .h5

In MATLAB, there is no high-level API access to manipulating group attributes that have already been created. See [this MATHWORKS question](https://www.mathworks.com/matlabcentral/answers/1860823-is-there-a-way-to-identify-group-names-in-an-h5-file-programmatically-without-using-h5info), [this blog post](http://jdstechjournal.blogspot.com/2015/11/renaming-hdf5-datasets-in-matlab.html).

In general, it is not possible to either delete a dataset or change its datatype. From section 5.3.2 of the [HDF5 manual](http://www.hdfgroup.org/HDF5/doc/UG/UG_frame10Datasets.html):

> The datatype is set when the dataset is created and can never be changed.

This is due to how space is assigned in an HDF5 file. While it's not possible to delete a dataset (for the same reasons), it can be "unlinked" and be made inaccessible, but this does not reclaim the used space.

There are 2 choices to change data-types: 
1. unlink the old dataset and create a new one in its place. 
- The new dataset can have the same name as the old one. 
2. Create an entirely new HDF5 file, and copy the old data into the new one.

Neither of these options reclaims used space so files can get bloat. For this reason, we should avoid storing multiple `datasets` in a file, where a dataset is any group (e.g. a time-series for a single z-plane) that we may want to process 

# .tiff
- [tifffile](https://github.com/cgohlke/tifffile/blob/master/tifffile/tifffile.py): Saving our data as a `.tiff` will require changing some metadata to account for any pixel trimming but is a good file format with most image processing libraries, may be worth doing in the future.
